layers/cmd/ingest_csv.go

```diff
@@ -195,8 +197,107 @@ func (ci *CSVIngester) Run(pipeline *pipeline.PipelineContext) error {
 	return nil
 }
 
+type CSVJoinIngester struct {
+	BaseIngestParams
+	StartRow     int    `json:"startRow" yaml:"startRow"`
+	EndRow       int    `json:"endRow" yaml:"endRow"`
+	HeaderRow    int    `json:"headerRow" yaml:"headerRow"`
+	ID           string `json:"id" yaml:"id"`
+	IngestByRows bool   `json:"ingestByRows" yaml:"ingestByRows"`
+	Delimiter    string `json:"delimiter" yaml:"delimiter"`
+	ColumnRange  uint   `json:"columnRange" yaml:"columnRange"`
+	initialized  bool
+	ingester     *ls.Ingester
+}
+
+type CSVJoinConfig struct {
+	VariantID        string
+	StartCol, EndCol int
+	IDCols           []int
+}
+
+type joinData struct {
+	CSVJoinConfig
+	data []string
+	id   string
+}
+
+func (cji *CSVJoinIngester) ingestCSVJoin(entities []CSVJoinConfig, stream io.ReadCloser) error {
+	reader := csv.NewReader(stream)
+	reader.Comma = '\t'
+	// reader.LazyQuotes = true
+	// reader.FieldsPerRecord = -1
+	seenIds := make(map[string][]string)
+	joinCtx := make([]joinData, 0)
+
+ROWS:
+	for row := 0; ; row++ {
+		rowData, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if row == cji.HeaderRow {
+			continue
+		}
+
+		// compare between joinCtx[0].id and first range of row
+		if len(joinCtx) > 0 {
+			checkRange := GenerateHashFromIDs(rowData[0:len(joinCtx[0].data)])
+			if checkRange != joinCtx[0].id {
+				// new graph, output previous
+				joinCtx = make([]joinData, 0)
+				seenIds = make(map[string][]string)
+				fmt.Println("new graph")
+			}
+		}
+		for _, schema := range entities {
+			if schema.EndCol < len(rowData) {
+				data := rowData[schema.StartCol : schema.EndCol+1]
+				hashID := GenerateHashFromIDs(data)
```

Comment:
Generate ID from the whole row data, not the sliced row data
GenerateHashFromIDs(rowData,ids)

Generate ID using ID columns
If no ID columns, use all columns